### PR TITLE
Add bound cidrs to tokens in AppRole

### PR DESCRIPTION
**Pull request by tyrannosaurus-becks**
AppRole already supports the notion of bound CIDRs, but it was only being used on login operations because we didn't yet support the ability to use them on tokens at the time it was added.

We now do, so this takes bound CIDRs one step further and only allows _tokens_ to be used by CIDRs in the given range.

To support both current users and this new functionality, we change the name of `"bound_cidr_list"` to `"secret_id_bound_cidrs"` to be more specific, and map the former parameter to the new name. We leave behind `"bound_cidr_list"` for backwards compatibility. 

We also add a new "`token_bound_cidrs"` parameter for those who would like all the previous features, _plus_ the ability require that the returned auth token is only used from a particular CIDR range.

### CODE REVIEW

File: builtin/credential/approle/path_login.go [hunk 1]
```diff
@@ -276,10 +282,16 @@ func (b *backend) pathLoginRenew(ctx context.Context, req *logical.Request, data
 		return nil, fmt.Errorf("role %q does not exist during renewal", roleName)
 	}
 
+	boundCIDRs, err := parseutil.ParseAddrs(role.BoundCIDRList)
```
[the remark this answers is not in the document]
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_login.go]: Oh, I didn't know that! I did the same with other renewals. I'll strip it out here, and pull it out in the other places too. Don't want to create cruft.

**Comment by tyrannosaurus-becks** [comment]
@jefferai yes, it does change the behavior. If somebody had implemented a workflow where one instance acquired a token, and then it was used by instances on a different cidr block, it would break that when they upgraded.

It seems like an edge case, but also something that statistically is likely to have been implemented by someone. Does it make more sense to be consistent with our use of the "bound_cidrs" parameter, and to document this in the release changelog? Or to create a different config parameter in this special case?

**Comment by jefferai** [comment]
I think let's document this in the CHANGES section and if people complain, we can consider reverting the behavior and adding a separate parameter instead. But it does seem like a reasonable thing to do, that we simply couldn't support before.

**Comment by jefferai** [comment]
Update -- @chrishoffman pointed out some reasons that people may want to have these be distinct, so I think they should be separate params.

**Comment by vishalnayak** [comment]
I am in favor of having separate parameters as well.

File: builtin/credential/approle/path_role.go [hunk 2]
```diff
@@ -948,16 +1021,17 @@ func (b *backend) pathRoleRead(ctx context.Context, req *logical.Request, data *
...
-		"policies":           role.Policies,
-		"secret_id_num_uses": role.SecretIDNumUses,
-		"secret_id_ttl":      role.SecretIDTTL / time.Second,
-		"token_max_ttl":      role.TokenMaxTTL / time.Second,
-		"token_num_uses":     role.TokenNumUses,
-		"token_ttl":          role.TokenTTL / time.Second,
-		"local_secret_ids":   false,
+		"bind_secret_id":        role.BindSecretID,
+		"secret_id_bound_cidrs": role.SecretIDBoundCIDRs,
+		"token_bound_cidrs":     role.TokenBoundCIDRs,
```
Comment (tyrannosaurus-becks) [inline builtin/credential/approle/path_role.go]: For discussion: replacing `bound_cidr_list` here with the two new parameters could break people's code. Should I return them all?

File: builtin/credential/approle/path_role.go [hunk 3]
```diff
@@ -1347,7 +1440,8 @@ func (b *backend) pathRoleBoundCIDRListRead(ctx context.Context, req *logical.Re
 	} else {
 		return &logical.Response{
 			Data: map[string]interface{}{
-				"bound_cidr_list": role.BoundCIDRList,
+				"secret_id_bound_cidrs": role.SecretIDBoundCIDRs,
+				"token_bound_cidrs":     role.TokenBoundCIDRs,
```
Comment (tyrannosaurus-becks) [inline builtin/credential/approle/path_role.go]: Same discussion as above.

File: builtin/credential/approle/path_role.go (hunk 2 above)
Comment (vishalnayak) [reply builtin/credential/approle/path_role.go]: Yeah, it will break backwards compatibility. I am leaning towards replicating `secret_id_bound_cidrs` in another field `bound_cidr_list` with an additional warning added to the response object that `bound_cidr_list` is deprecated and will not be returned in future releases. (and a note in code as a reminder for us to remove it).

File: builtin/credential/approle/validation.go [hunk 4]
```diff
@@ -68,18 +68,25 @@ type secretIDAccessorStorageEntry struct {
...
 
+	// If there are no CIDR blocks on the role, then the subset
+	// requirement would be satisfied
+	if len(role.SecretIDBoundCIDRs) != 0 {
+		subset, err := cidrutil.SubsetBlocks(role.SecretIDBoundCIDRs, secretIDCIDRs)
+		if !subset || err != nil {
+			return errwrap.Wrapf(fmt.Sprintf("failed to verify subset relationship between CIDR blocks on the role %q and CIDR blocks on the secret ID %q: {{err}}", role.SecretIDBoundCIDRs, secretIDCIDRs), err)
+		}
+	}
+	if len(role.TokenBoundCIDRs) != 0 {
```
Comment (vishalnayak) [inline builtin/credential/approle/validation.go]: I wonder if we really need this check. We are only ever going to pass in only the secret ID CIDRs to this function for subset checking. Consider that secret IDs are generated by a orchestration module sitting under a different CIDR than the nodes that will be _using_ the tokens belonging to a disjoint CIDR block. Then essentially what we are asserting here is that both during login time and when updating secret IDs, the secret ID CIDRs to be a subset of TokenBoundCIDRs.

I am thinking that we are better off leaving the TokenBoundCIDRs out of this function.

File: builtin/credential/approle/path_role.go (hunk 3 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: As discussed above, will return the deprecated param here and a warning.

File: builtin/credential/approle/validation.go (hunk 4 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/validation.go]: Ah! Thank you!

File: website/source/api/auth/approle/index.html.md [hunk 5]
```diff
@@ -69,9 +69,17 @@ enabled while creating or updating a role.
...
 - `bind_secret_id` `(bool: true)` - Require `secret_id` to be presented when
   logging in using this AppRole.
-- `bound_cidr_list` `(array: [])` - Comma-separated string or list of CIDR
+- `bound_cidr_list` `(array: [])` - Deprecated: Comma-separated string or list of CIDR
   blocks; if set, specifies blocks of IP addresses which can perform the login
   operation.
+- `secret_id_bound_cidrs` `(array: [])` - Comma-separated string or list of CIDR
+  blocks; if set, specifies blocks of IP addresses which can perform the login
+  operation.
+- `token_bound_cidrs` `(array: [])` - Comma-separated string or list of CIDR
```
Comment (vishalnayak) [inline website/source/api/auth/approle/index.html.md]: We might want to keep `token_bound_cidrs` to behave as just that and not mix it up with `secret_id_bound_cidrs`. It is less confusing that way and becomes intuitive by their names.

File: builtin/credential/approle/path_role.go [hunk 6]
```diff
@@ -1314,17 +1360,35 @@ func (b *backend) pathRoleBoundCIDRListUpdate(ctx context.Context, req *logical.
 		return nil, nil
 	}
 
-	role.BoundCIDRList = data.Get("bound_cidr_list").([]string)
-	if len(role.BoundCIDRList) == 0 {
-		return logical.ErrorResponse("missing bound_cidr_list"), nil
+	if cidrListRaw, ok := data.GetOk("secret_id_bound_cidrs"); ok {
```
[the remark this answers is not in the document]
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: Yes, I agree, it does make the code much more complex than I initially expected. I'm going to add the ability to do all three of those fields, and then let everyone see what this looks like IRL to decide whether this is really the way we want to go.

File: website/source/api/auth/approle/index.html.md (hunk 5 above)
Comment (tyrannosaurus-becks) [reply website/source/api/auth/approle/index.html.md]: Yes, I thought of that approach too. I'm happy to go either way. The reason I didn't pick it initially is because, then the user would have to specify bound CIDRs in both parameters. I thought that could be clunky to use. Also, I _think_ they'd provide identical CIDRs in each list, and that it would work fine even if they didn't, but it would be worthwhile to carefully think through scenarios and perhaps add tests.

On the other hand, in the current scenario they could already provide differing CIDR blocks in each parameter. :-)

File: builtin/credential/approle/path_login.go [hunk 7]
```diff
@@ -219,17 +220,30 @@ func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, dat
...
-		belongs, err := cidrutil.IPBelongsToCIDRBlocksSlice(req.Connection.RemoteAddr, role.BoundCIDRList)
-		if err != nil || !belongs {
-			return logical.ErrorResponse(errwrap.Wrapf(fmt.Sprintf("source address %q unauthorized by CIDR restrictions on the role: {{err}}", req.Connection.RemoteAddr), err).Error()), nil
+		if len(role.SecretIDBoundCIDRs) != 0 {
+			belongs, err := cidrutil.IPBelongsToCIDRBlocksSlice(req.Connection.RemoteAddr, role.SecretIDBoundCIDRs)
+			if err != nil || !belongs {
+				return logical.ErrorResponse(errwrap.Wrapf(fmt.Sprintf("source address %q unauthorized by CIDR restrictions on the role: {{err}}", req.Connection.RemoteAddr), err).Error()), nil
+			}
+		}
+		if len(role.TokenBoundCIDRs) != 0 {
```
Comment (jefferai) [inline builtin/credential/approle/path_login.go]: We shouldn't be doing this check -- login should check against secret id bound CIDRs only, not both.

File: builtin/credential/approle/path_role.go [hunk 8]
```diff
@@ -749,11 +778,17 @@ func (b *backend) roleEntry(ctx context.Context, s logical.Storage, roleName str
...
 	if role.BoundCIDRListOld != "" {
-		role.BoundCIDRList = strings.Split(role.BoundCIDRListOld, ",")
+		role.SecretIDBoundCIDRs = strings.Split(role.BoundCIDRListOld, ",")
 		role.BoundCIDRListOld = ""
 		needsUpgrade = true
 	}
 
+	if len(role.BoundCIDRList) != 0 {
+		role.SecretIDBoundCIDRs = role.BoundCIDRList
+		role.BoundCIDRList = []string{}
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Can just set to nil

File: builtin/credential/approle/path_role.go [hunk 9]
```diff
@@ -840,14 +875,28 @@ func (b *backend) pathRoleCreateUpdate(ctx context.Context, req *logical.Request
...
+		valid, err := cidrutil.ValidateCIDRListSlice(role.SecretIDBoundCIDRs)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to validate CIDR blocks: {{err}}", err)
+		}
+		if !valid {
+			return logical.ErrorResponse("invalid CIDR blocks"), nil
+		}
+	}
+
+	if boundCIDRListRaw, ok := data.GetOk("token_bound_cidrs"); ok {
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: These three lines can just be `role.TokenBoundCIDRs = data.Get("token_bound_cidrs").([]string)`

File: builtin/credential/approle/path_role.go [hunk 10]
```diff
@@ -1314,17 +1368,38 @@ func (b *backend) pathRoleBoundCIDRListUpdate(ctx context.Context, req *logical.
...
+	if cidrListRaw, ok := data.GetOk("secret_id_bound_cidrs"); ok {
+		role.SecretIDBoundCIDRs = cidrListRaw.([]string)
+	} else if cidrListRaw, ok := data.GetOk("bound_cidr_list"); ok {
+		role.SecretIDBoundCIDRs = cidrListRaw.([]string)
 	}
 
-	valid, err := cidrutil.ValidateCIDRListSlice(role.BoundCIDRList)
-	if err != nil {
-		return nil, errwrap.Wrapf("failed to validate CIDR blocks: {{err}}", err)
+	if cidrListRaw, ok := data.GetOk("token_bound_cidrs"); ok {
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Same here.

File: builtin/credential/approle/path_role.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jefferai) [reply builtin/credential/approle/path_role.go]: I think this API should do exactly what it did before, and simply have two new APIs defined. I don't love overloading all of the cidr handling here. I'm also not sure if it's necessary, so I'd be fine with simply reverting the changes to this function (other than storing the values at the new struct location) and not adding new broken-out API functions unless it becomes something people care about.

File: website/source/api/auth/approle/index.html.md (hunk 5 above)
Comment (jefferai) [reply website/source/api/auth/approle/index.html.md]: I'm much more happy with users specifying the same CIDRs in both parameters than with the secret ID value doing double-duty. Also, in many cases where users might have put the same exact values, users will only need to specify token cidrs; secret id cidrs end up simply preventing a token that can't be used from being created and taking up a lease.

File: builtin/credential/approle/path_role.go (hunk 9 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: I thought that initially too, but it turned out that if I did that, then on the update operation it wiped out any previously set values if that field wasn't explicitly set by the caller. So I added the ok check so that I'd only overwrite the value if it was actually sent to me.

File: builtin/credential/approle/path_role.go (hunk 10 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: Same thing - if I do that and an update operation only changes _other_ fields, it wipes this out.

File: builtin/credential/approle/path_login.go (hunk 7 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_login.go]: :heavy_check_mark:

File: builtin/credential/approle/path_role.go (hunk 8 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: :heavy_check_mark:

File: builtin/credential/approle/path_role.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: :heavy_check_mark:

File: website/source/api/auth/approle/index.html.md (hunk 5 above)
Comment (tyrannosaurus-becks) [reply website/source/api/auth/approle/index.html.md]: Cool cool. Will update.

Comment (tyrannosaurus-becks) [reply website/source/api/auth/approle/index.html.md]: @jefferai have pushed commits incorporating all feedback so far.

File: builtin/credential/approle/path_role.go (hunk 9 above)
Comment (jefferai) [reply builtin/credential/approle/path_role.go]: Oh, right, I forgot it's create/update, not just update! Sounds good.

File: builtin/credential/approle/path_role.go (hunk 10 above)
Comment (jefferai) [reply builtin/credential/approle/path_role.go]: Yep, keep as-is.

File: website/source/api/auth/approle/index.html.md (hunk 5 above)
Comment (jefferai) [reply website/source/api/auth/approle/index.html.md]: I still see all three values being set via bound_cidr_list...

File: builtin/credential/approle/path_role.go [hunk 11]
```diff
@@ -1377,6 +1522,54 @@ func (b *backend) pathRoleBoundCIDRListDelete(ctx context.Context, req *logical.
 	return nil, b.setRoleEntry(ctx, req.Storage, roleName, role, "")
 }
 
+func (b *backend) pathRoleSecretIDBoundCIDRDelete(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
```
Comment (kalafut) [inline builtin/credential/approle/path_role.go]: This function and the next are nearly identical. Maybe you can extract the common core into a parameterized version.

File: builtin/credential/approle/path_role.go [hunk 12]
```diff
@@ -1314,12 +1400,21 @@ func (b *backend) pathRoleBoundCIDRListUpdate(ctx context.Context, req *logical.
 		return nil, nil
 	}
 
-	role.BoundCIDRList = data.Get("bound_cidr_list").([]string)
-	if len(role.BoundCIDRList) == 0 {
+	var cidrs []string
+	if cidrsIfc, ok := data.GetOk("bound_cidr_list"); ok {
+		cidrs = cidrsIfc.([]string)
+		role.BoundCIDRList = cidrs
```
Comment (vishalnayak) [inline builtin/credential/approle/path_role.go]: Can we prioritize `secret_id_bound_cidrs` as the one that gets checked before `bound_cidr_list`?

Also, should we really set the value into `role.BoundCIDRList`? Can we store it in `role.SecretIDBoundCIDRs` field instead? This can avoid an upgrade upon the role read.

File: builtin/credential/approle/path_role_test.go [hunk 13]
```diff
@@ -1121,14 +1121,491 @@ func TestAppRole_RoleCRUD(t *testing.T) {
 	var err error
 	b, storage := createBackendWithStorage(t)
 
+	roleData := map[string]interface{}{
```
Comment (vishalnayak) [inline builtin/credential/approle/path_role_test.go]: Unless I am reading this diff wrong, the tests are for RUD of all the other fields but the fields that are getting modified in this PR. I assume that you wanted to the that immediately follows this to do the testing of the new fields. But that has test steps for all the other fields as well. Am I missing something?

File: website/source/api/auth/approle/index.html.md [hunk 14]
```diff
@@ -69,9 +69,15 @@ enabled while creating or updating a role.
 - `role_name` `(string: <required>)` - Name of the AppRole.
 - `bind_secret_id` `(bool: true)` - Require `secret_id` to be presented when
   logging in using this AppRole.
-- `bound_cidr_list` `(array: [])` - Comma-separated string or list of CIDR
+- `bound_cidr_list` `(array: [])` - Deprecated: Comma-separated string or list of CIDR
```
Comment (vishalnayak) [inline website/source/api/auth/approle/index.html.md]: I would just remove this field from the docs. Those who are using this will get a warning in the response. But those who are not already using it, will naturally start using the new one.

File: builtin/credential/approle/path_role.go (hunk 11 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: Hah, for some reason yesterday I didn't think so but looking at it this morning it's pretty obvious it'd be easy to do. :-) Will update.

File: builtin/credential/approle/path_role.go [hunk 15]
```diff
@@ -51,12 +51,18 @@ type roleStorageEntry struct {
...
-	// A constraint, if set, specifies the CIDR blocks from which logins should be allowed
+	// Deprecated: A constraint, if set, specifies the CIDR blocks from which logins should be allowed
 	BoundCIDRListOld string `json:"bound_cidr_list,omitempty"`
 
-	// A constraint, if set, specifies the CIDR blocks from which logins should be allowed
+	// Deprecated: A constraint, if set, specifies the CIDR blocks from which logins should be allowed
 	BoundCIDRList []string `json:"bound_cidr_list_list" structs:"bound_cidr_list" mapstructure:"bound_cidr_list"`
 
+	// A constraint, if set, specifies the CIDR blocks from which logins should be allowed
+	SecretIDBoundCIDRs []string `json:"secret_id_bound_cidrs" structs:"bound_cidr_list" mapstructure:"bound_cidr_list"`
```
[the remark this answers is not in the document]
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: I did that because the `BoundCIDRList` above is sort of funny. Its json tag is `bound_cidr_list_list`, but the rest of its tags are `bound_cidr_list`. I thought perhaps to be safe, I should use the same mapping and struct tags as it. 

I grepped the package and couldn't find any use of `structs.New(role)` or `structs.Map(role)`, so I was tempted to strip all the `struct` and `mapstructure` tags altogether, but I didn't want to go too far. Though `roleStorageEntry` isn't exported, I wondered if someone might have a fork somewhere where I shouldn't mess with how it's sent to struct/map. 

TLDR; I was thrown by the inconsistent `BoundCIDRLIst` tags. :-)

File: builtin/credential/approle/path_role.go [hunk 16]
```diff
@@ -121,10 +127,21 @@ func rolePaths(b *backend) []*framework.Path {
 					Default:     true,
 					Description: "Impose secret_id to be presented when logging in using this role. Defaults to 'true'.",
 				},
+				// Deprecated
 				"bound_cidr_list": &framework.FieldSchema{
+					Type: framework.TypeCommaStringSlice,
+					Description: `Deprecated: Comma separated string or list of CIDR blocks. If set, specifies the blocks of
+IP addresses which can perform the login operation.`,
```
[the remark this answers is not in the document]
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: :+1:

File: builtin/credential/approle/path_role.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (chrishoffman) [reply builtin/credential/approle/path_role.go]: Yeah, I omit the structs and mapstructure tags by default at this point.  They probably aren't necessary and can likely be removed.

File: builtin/credential/approle/path_role.go [hunk 17]
```diff
@@ -749,11 +810,17 @@ func (b *backend) roleEntry(ctx context.Context, s logical.Storage, roleName str
 	needsUpgrade := false
 
 	if role.BoundCIDRListOld != "" {
-		role.BoundCIDRList = strings.Split(role.BoundCIDRListOld, ",")
+		role.SecretIDBoundCIDRs = strings.Split(role.BoundCIDRListOld, ",")
```
[the remark this answers is not in the document]
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: That's handy.

File: builtin/credential/approle/path_role.go (hunk 12 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: Ah! You just reminded me that I was also going to write a `fieldData.GetFirst()` method. Yes, I'll put the result directly into `role.SecretIDBoundCIDRs`.

File: builtin/credential/approle/path_role_test.go (hunk 13 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role_test.go]: Maybe? This test is identical to the RUD one above, but instead of looking for "bound_cidr_list", it covers "secret_id_bound_cidrs" at the role level. I wanted to make sure the previous test worked, and that the new field worked in an identical fashion. You're right though that I didn't add update tests for the new ones, and I should add that.

File: website/source/api/auth/approle/index.html.md (hunk 14 above)
Comment (tyrannosaurus-becks) [reply website/source/api/auth/approle/index.html.md]: Ah! Good idea.

File: builtin/credential/approle/path_role.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: Huh, I just merged in master and someone when through and stripped the struct tags from this object, so I did too. Also updated my mapstructure tag to match the json tag.

File: builtin/credential/approle/path_role.go [hunk 18]
```diff
@@ -978,6 +1064,7 @@ func (b *backend) pathRoleRead(ctx context.Context, req *logical.Request, data *
 	if err := validateRoleConstraints(role); err != nil {
 		resp.AddWarning("Role does not have any constraints set on it. Updates to this role will require a constraint to be set")
 	}
+	resp.AddWarning(`The "bound_cidr_list" parameter is deprecated and will be removed. Please use "secret_id_bound_cidrs" instead.`)
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Shouldn't this have an if condition so it's only added if they actually set `bound_cidr_list`? Probably this should go up above where you're actually looking at the parameters (although it would rather negate the point of using GetFirst since you'd have to check if bound_cidr_list is set anyways).

File: builtin/credential/approle/path_role.go [hunk 19]
```diff
@@ -1363,16 +1468,36 @@ func (b *backend) pathRoleBoundCIDRListRead(ctx context.Context, req *logical.Re
...
+			resp := &logical.Response{
+				Data: map[string]interface{}{
+					"bound_cidr_list": role.BoundCIDRList,
+				},
+			}
+			resp.AddWarning(`The "bound_cidr_list" parameter is deprecated and will be removed. Please use "secret_id_bound_cidrs" instead.`)
+			return resp, nil
+		default:
+			// shouldn't occur IRL
+			return nil, errors.New("unrecognized field provided: " + fieldName)
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Is there a reason you're pulling in the `go-errors` package instead of the stdlib `errors`?

Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: Arg, misbehaving code editor. I'll fix.

File: builtin/credential/approle/path_role.go (hunk 18 above)
Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: That would be a somewhat significant code change to make because, if they're using `bound_cidr_list`, the first time the role is read by the `roleEntry` method, if they're using that field it gets moved and cleared. So we'd only know on first read, and it would take some refactoring to move that information out of the `roleEntry` method. 

Another options would be to strip the warning. Or we could leave it there for all read requests during the deprecation period.

Comment (jefferai) [reply builtin/credential/approle/path_role.go]: I think just remove it. I'd rather not return a warning if they didn't actually set it. Chances are that code to handle the old parameter will be there forever anyways :-)

Comment (jefferai) [reply builtin/credential/approle/path_role.go]: After checking out the branch and looking more at the code (outside of just a diff context) it was more clear why this warning was here. I changed the wording slightly, and I think we can leave it in for a bit and then remove both it and the return value.

Comment (tyrannosaurus-becks) [reply builtin/credential/approle/path_role.go]: Thank you!

Comment (jefferai) [reply builtin/credential/approle/path_role.go]: Thank _you_ for carrying this through!
